### ISSUE TITLE
undefined reference to `_nms'

### ISSUE THREAD

**Issue by nasica88** [laonbud/SoooA#3]
While building SoooA on ppc64le, I encountered the following error.

Building target: SoooaServerDebug
Invoking: NVCC Linker
nvcc --cudart static -Xlinker --export-dynamic --relocatable-device-code=false -gencode arch=compute_60,code=sm_60 -link -o "SoooaServerDebug"  ./src/main_jp.o ./src/main.o ./src/ssd/MultiBoxLossLayer.o ./src/ssd/BBoxUtil.o ./src/ssd/DetectionOutputLayer.o ./src/ssd/ssd_common.o ./src/ssd/PriorBoxLayer.o ./src/ssd/AnnotationDataLayer.o ./src/ssd/ConcatLayer.o ./src/ssd/AnnotatedDataLayer.o ./src/ssd/NormalizeLayer.o ./src/ssd/PermuteLayer.o ./src/ssd/FlattenLayer.o ./src/ssd/DetectionEvaluateLayer.o ./src/util/DebugUtil.o ./src/util/Util.o ./src/util/ImageCropper.o ./src/util/ImagePacker.o ./src/util/MathFunctions.o ./src/util/CudaUtils.o ./src/util/Sampler.o ./src/util/ImageUtil.o ./src/util/ConfusionMatrix.o ./src/util/RNG.o ./src/dataset/SDF.o ./src/dataset/MockDataSet.o ./src/dataset/ImTransforms.o ./src/dataset/DataSet.o ./src/dataset/ImagePackDataSet.o ./src/dataset/DataReader.o ./src/dataset/IO.o ./src/dataset/DataTransformer.o ./src/client/Client.o ./src/client/ClientAPI.o ./src/client/test/RunNetworkTest.o ./src/client/test/CreateNetworkTest.o ./src/client/test/PrintMeasureTest.o ./src/client/test/SaveLoadNetworkTest.o ./src/client/test/RunNetworkWithInputTest.o ./src/prop/EnumDef.o ./src/prop/LayerFunc.o ./src/prop/PropMgmt.o ./src/prop/LayerPropList.o ./src/prop/InnerLayerFunc.o ./src/prop/NetworkProp.o ./src/prop/test/PropTest.o ./src/measure/MeasureEntry.o ./src/measure/MeasureManager.o ./src/frcnn/FrcnnTestVideoOutputLayer.o ./src/frcnn/RoIInputLayer.o ./src/frcnn/RoITestVideoInputLayer.o ./src/frcnn/ProposalLayer.o ./src/frcnn/FrcnnTestLiveOutputLayer.o ./src/frcnn/RoITestInputLayer.o ./src/frcnn/RoITestLiveInputLayer.o ./src/frcnn/ProposalTargetLayer.o ./src/frcnn/FrcnnTestOutputLayer.o ./src/frcnn/AnchorTargetLayer.o ./src/frcnn/RoIPoolingLayer.o ./src/tools/ParamManipulator.o ./src/tools/Tools.o ./src/log/HotCode.o ./src/log/ColdLog.o ./src/log/SysLog.o ./src/log/HotLogContext.o ./src/log/StdOutLog.o ./src/log/HotLog.o ./src/tensor/SyncMem.o ./src/tensor/SyncMem_device.o ./src/tensor/Data.o ./src/perf/PerfList.o ./src/perf/Perf.o ./src/cuda/Cuda.o ./src/common/FileMgmt.o ./src/common/common.o ./src/common/test.o ./src/application/DeepDream.o ./src/application/ArtisticStyle.o ./src/application/ApplicationCudaFunctions.o ./src/plan/PlanOptimizer.o ./src/plan/PlanBuilder.o ./src/plan/PlanParser.o ./src/plan/PhysicalPlan.o ./src/plan/LogicalPlan.o ./src/plan/ResourceManager.o ./src/plan/test/PlanOptimizerTest.o ./src/plan/test/PlanBuilderTest.o ./src/plan/test/PlanParserTest.o ./src/test/TestMain.o ./src/test/TestUtil.o ./src/network/Network.o ./src/network/Donator.o ./src/network/test/NetworkRunByPlanTest.o ./src/updater/Update_device.o ./src/updater/Update.o ./src/updater/Updater.o ./src/layer/RecallLayer.o ./src/layer/ElementWiseLayer.o ./src/layer/FullyConnectedLayer_device.o ./src/layer/SplitLayer.o ./src/layer/FullyConnectedLayer.o ./src/layer/PoolingLayer_device.o ./src/layer/NoiseInputLayer.o ./src/layer/YOLOLossLayer_device.o ./src/layer/CelebAInputLayer.o ./src/layer/BatchNormLayer.o ./src/layer/PoolingLayer.o ./src/layer/BatchNormLayer_device.o ./src/layer/MultiLabelDataInputLayer.o ./src/layer/SoftmaxLayer.o ./src/layer/ReluLayer.o ./src/layer/BaseLayer.o ./src/layer/LearnableLayer.o ./src/layer/Sigmoid2Layer_device.o ./src/layer/InterpolationLayer.o ./src/layer/DataInputLayer.o ./src/layer/CustomInputLayer.o ./src/layer/InputLayer.o ./src/layer/CrossEntropyWithLossLayer_device.o ./src/layer/VOCPascalInputLayer.o ./src/layer/AccuracyLayer.o ./src/layer/SilenceLayer.o ./src/layer/ReshapeLayer.o ./src/layer/ILSVRCInputLayer.o ./src/layer/LRNLayer.o ./src/layer/DepthConcatLayer.o ./src/layer/LRNLayer_device.o ./src/layer/SegAccuracyLayer.o ./src/layer/SoftmaxWithLossLayer_device.o ./src/layer/DropOutLayer.o ./src/layer/KistiInputLayer.o ./src/layer/HyperTangentLayer_device.o ./src/layer/ConvLayer_device.o ./src/layer/DepthConcatLayer.o ./src/layer/SmoothL1LossLayer.o ./src/layer/DummyLossLayer.o ./src/layer/ReluLayer_device.o ./src/layer/DropOutLayer_device.o ./src/layer/ConvLayer.o ./src/layer/SoftmaxWithLossLayer.o ./src/layer/ImageSegDataLayer.o ./src/layer/DummyInputLayer.o ./src/layer/test/CustomInputTest.o ./src/param/InitParam.o ./src/param/ParamMgmt.o ./src/param/Param.o ./src/capi/CapiToolAPI.o ./src/capi/CapiClientAPI.o ./src/sess/MsgSerializer.o ./src/sess/Communicator.o ./src/3rd_party/tinyxml2/tinyxml2.o ./src/3rd_party/nms/nms_kernel.o ./src/3rd_party/nms/gpu_nms.o ./src/3rd_party/nms/cpu_nms.o ./src/3rd_party/jsoncpp/jsoncpp.o ./src/3rd_party/cnpy/cnpy.o ./src/core/ThreadMgmt.o ./src/core/Task.o ./src/core/Worker.o ./src/core/InputDataProvider.o ./src/core/Broker.o ./src/core/Receiver.o ./src/core/Sender.o ./src/core/Job.o ./src/core/WorkContext.o ./src/core/test/BrokerTest.o ./src/examples/KistiGan.o ./src/examples/YOLO.o ./src/examples/Examples.o ./src/examples/frcnn/FRCNN.o ./src/examples/GAN/GAN.o ./src/examples/KISTIKeyword/KISTIKeyword.o ./src/examples/VGG16/VGG16.o ./src/examples/SSD/SSD.o ./src/examples/LeNet/LeNet.o ./src/monitor/LossConsole.o  -lcudart -lopencv_core -lcublas -lcudnn -lboost_system -lboost_filesystem -lopencv_highgui -lopencv_imgproc -lopencv_imgcodecs -lopencv_features2d -lopencv_video -lopencv_videoio -lz -lboost_iostreams -lpython2.7 -lX11 -lgpu_nms -lboost_serialization -lboost_thread -lboost_python -L/usr/local/boost/lib -L../src/3rd_party/nms -L/usr/local/lib -L/usr/lib/powerpc64le-linux-gnu
/usr/bin/ld: warning: libboost_system.so.1.58.0, needed by //usr/lib/powerpc64le-linux-gnu/libmirclient.so.9, may conflict with libboost_system.so.1.65.1
/usr/bin/ld: warning: libboost_filesystem.so.1.58.0, needed by //usr/lib/powerpc64le-linux-gnu/libmircommon.so.7, may conflict with libboost_filesystem.so.1.65.1
./src/3rd_party/nms/gpu_nms.o: In function `__pyx_pf_3nms_7gpu_nms_gpu_nms':
**/home/u0017649/SoooA/SoooA/DebugGen/../src/3rd_party/nms/gpu_nms.c:1928: undefined reference to `_nms'**
collect2: error: ld returned 1 exit status
makefile:93: recipe for target 'SoooaServerDebug' failed
make: *** [SoooaServerDebug] Error 1
ERROR: build stopped

Since your git repository contained binary *.so files for x86 only, I built those from py-faster-rcnn source as below, and copied the generated gpu_nms.so to the ~/SoooA/SoooA/src/3rd_party/nms directory, as gpu_nms.so and libgpu_nms.so.

u0017649@sys-90725:~/py-faster-rcnn/lib$ make

u0017649@sys-90725:~/SoooA/SoooA/src/3rd_party/nms$ file gpu_nms.so libgpu_nms.so
gpu_nms.so:    ELF 64-bit LSB shared object, 64-bit PowerPC or cisco 7500, version 1 (SYSV), dynamically linked, BuildID[sha1]=fcd4352ddb650b5e83b795572d54d6f278a0b001, not stripped
libgpu_nms.so: ELF 64-bit LSB shared object, 64-bit PowerPC or cisco 7500, version 1 (SYSV), dynamically linked, BuildID[sha1]=fcd4352ddb650b5e83b795572d54d6f278a0b001, not stripped

Is this not correct way to build libgpu_nms.so ?

**Comment by whatisain** [comment]
We tested SoooA framework only on x86 linux environment since we do not have other environments like PowerPC.

If you did same as follows, you took the right action.

(1) clone py-faster-rcnn codes from https://github.com/rbgirshick/py-faster-rcnn.git
(2) move to py-faster-rcnn/lib directory
(3) make
(4) copy py-faster-rcnn/lib/nms/gpu_nms.so to $SOOOA_HOME/src/3rd_party/nms/libgpu_nms.so

We tested above process again and confirmed that it works well on x86 machine.

Just for your information, 
"SoooA" is old project and deprecated. We moved to "LaonSill" and it is most up-to-date.
If you still have build issues, we suggest you to try "LaonSill".

You can clone "LaonSill" project from
https://github.com/laonbud/LaonSill

Thanks
